### DISCUSSION TITLE
Any Helm chart designed to deploy the Kargo CRDs dynamically per-application?

### DISCUSSION THREAD

**Discussion by pww217** [akuity/kargo#2472] [Ideas]
(Saw a few discussions tangental to this but not directly about it, although [this one](https://github.com/akuity/kargo/discussions/1357) is pretty close)

I just set up Kargo using the QuickStart example on my local machine and am very impressed. I am hoping it can help us deal with the spaghetti code we often have in GHA to bridge the gap between the CI and ArgoCD like promotions, integration testing, and more. Familiar enough use case to all here, I'm sure.

I am however having a lot of trouble figuring out how to apply Kargo to existing applications, as all examples so far basically just put a series of direct manifests and apply them, [like so.](https://github.com/krancour/kargo-examples/blob/main/02-git-driven/02-helm-driven/02-commit-n-image/kargo.yaml) as well as the QuickStart example. Notably, the linked example is likely the mechanism we'll be using (Helm + Git Repos + ECR) and I thank @krancour for the very helpful examples.

But because we have many apps and product - and I don't want to copy-paste these manifests a dozen times - I need something more scalable, and this is exactly what Helm is designed for after all right? We can template out the `Warehouse`, `Project`, and `Stage` resources can we not?

**So the actual question is**: Does such a helm chart exist to template out the CRDs per application, or what kind of pattern are y'all using in reality to reduce the toil there? Is there a solution anyone is willing to share?

The pattern I see in examples is to have all these CRDs living in a namespace on the cluster and then applying/gluing them to existing ArgoCD apps. That model should work well for me, but I am looking for a way to do it with a file of a few dozen or hundreds of lines, not thousands and thousands of lines of CRD manifests.

Please do let me know if I've got anything very wrong here. Kargo is pretty new to me but it seems very promising and I just want to understand how to scale it beyond these examples.

**Comment by krancour** [comment]
Hi @pww217 there has been discussion of eventually introducing high-level constructs like "StageSets" to manage many similar Stages. You could think of this as being similar to the way that Argo CD ApplicationSets can mangage many similar Applications.

But you are also correct to that Helm is an ideal tool for templating out your own patterns, which you probably don't need to repeat only from Stage to Stage within a Project, but probably will want to also repeat _across many Projects as well._

We do know of several users who are doing exactly that, so you are on the right track with this idea.

**Reply by pww217** [reply]
Perfect, thank you. It seems any such charts are not being offered publicly, just the baseline control-plane chart (which is quite good).

I'll probably try to think of a way to do this that can be pretty easily-readable and developer-friendly and if it works well, I may open it up as a public repo for those with similar use-cases. I imagine others will do something similar if they have, like us, a spoke-and-hub ArgoCD/multi-cluster model.

**Reply by krancour** [reply]
@pww217 as I think about this more, I am less certain that I fully understood your question or the chart you are proposing to create...

Are you trying to make it easier to manage many Projects with similar pipeline composition and promotion workflows?

Or are you looking for a chart that manages Kargo controllers that aren't part of Kargo's control plane?

**Reply by pww217** [reply]
>Are you trying to make it easier to manage many Projects with similar pipeline composition and promotion workflows?

I think it's this one. We have many products/apps and most run along very similar pipelines, and almost all are ArgoCD managed. They run across a few clusters. The CI/CD has up until now been done by some real spaghetti-code Github Actions and that's what I'm looking to replace with something clean, standardized (that's a big one), and easy to use like Kargo.

I want to replace everything GHA does except build/push images. So that's promotions, rollbacks, automated testing, canary rollouts, and more.

So I'm talking about a chart that would probably be structured somewhat like:

```
templates/
  stages.yaml
  warehouses.yaml
  analysistemplate.yaml
Chart.yaml
values.yaml
```
Maybe there'd be a `project.yaml` template too - still trying to figure out how to sort these into "buckets" in a way that both we platform engineers and our developers can understand.

The idea being that I can then deploy this chart with a short values file - probably < 20 lines for a very simple app, instead of > 100 lines of manifest for the same simple app. That values file can probably live in the app/developer's repo for them to manage, just like their app Helm values/configuration. 

The name of the game here for me is scalability. So far instance, realizing that it seems each Kargo Project may require its own namespace is not ideal. It livable but we have a lot of applications and didn't necessarily want 100 namespaces in our control cluster.

Seemingly related to the above is that it seems like the Kargo CR's should be in the same namespace as the Application, at least in the quickstarts and examples this is the case. Trying to figure out if that's a requirement, but at the moment I'm getting an error related to a mutating webhook when trying to sync an app that's in another namespace.

I imagine this is for least-privileges and security reasons, but since those apps/namespaces live in other clusters I'm not too sure how to make this work. Anyway sorry, a bit off topic. Just trying to give you some idea of where I am with all this.

**Reply by pww217** [reply]
But I wouldn't think it would have to be the same namespace, since ultimately what Kargo is trying to do is to alter the ArgoCD `Application` resource, which is on the same control cluster living in the `argocd` namespace, which makes this most recent promotion error all the more confusing:

```
Promotion Errored
error executing Argo CD promotion mechanism: Argo CD Application "some-app-us-east-1" in namespace "argocd" does not permit mutation by Kargo Stage dev in namespace kargo-some-app-admin
```

I figure there must be some issue with serviceAccount, roleBinding or similar. Anyway, I am trying to write down what I learn in this process and diagram it out a little. If it's useful I may contribute those upstream too at some point as the docs could definitely use some fleshing out.

Edit: Oh I think perhaps I'm just missing a required annotation, looking back at the Quickstart

```
      annotations:
        kargo.akuity.io/authorized-stage: kargo-demo:{{stage}}
```
So I'll try that out.

Edit2: And that fixed it! First time I've had Kargo work in a production-like env! Very exciting stuff.

**Reply by alexvojnovic** [reply]
@pww217 Were you ever successful in creating that Helm chart for streamlining creation of Kargo CRD's? We are currently trying to accomplish the same goal that you have discussed above. Would you be comfortable sharing the base chart if you were able to get a working example together?

**Reply by pww217** [reply]
Hey @alexvojnovic, unfortunately what I built ended up being pretty opinionated and never fully deployed. 

It was opinionated largely because the pipelines hadn't yet been put into a separate CRD. I was waiting for that to be the case so that it could be more flexible and generic, but I never got around to it and now I'm at a new employer entirely.

I believe it was the `PromotionTask` CRD I was waiting for in 1.2 or so, but I never actually adjusted the chart to use it. Still, the overall idea was to have this generic chart that would define the major components, but the "specifics" unique to your org (do you use helm, kustomize, ECR, dockerhub, etc. etc.) could be contained within this CRD and not have to be templated out in the chart.

If it's helpful at all, I can make it public somewhere and remove anything specific to my former company. I think there's a lot of promise for a chart like this given Kargo's current architecture which is far more modular than when I began working on it probably oof...1 to 1.5 years ago?

**Reply by pww217** [reply]
@alexvojnovic 

You inspired me a bit to give it another look. I don't really think it's a ridiculous reach to update it to the current Kargo architecture.

I removed any identifying info and pushed it to https://github.com/pww217/kargo-pipeline-helm

I had a couple conversations with cursor which is what generated the documents that outline current state, ideal state, and just clarified major resource types.

I may actually put some work into it, but I don't really have a sizable K8s cluster to test it again anymore, yet along several, just a small homelab (K3s) one.

But if you want to take it and run it it, or just get any inspiration at all from it, I'd be happy with that. I hope what I was trying to do is clear enough with the extra docs.
